### DISCUSSION TITLE
Cloze plugin improvements: RT and ACC!

### DISCUSSION THREAD

**Discussion by gustavoestivalet** [jspsych/jsPsych#919] [Show and tell]
Hi Guys,

Last week I did  a cloze experiment using the "jspsych-cloze" plugin. It was really easy to do the experiment using the "%%" tag, pretty smart! I figured out 4 amelioration:

1) There is no RT counter (I implemented it in the code below! Please, check it out!)

2) If the "check_answers" parameter is TRUE and the clozes are wrong, participant get blocked! I would be interesting to have an extra parameter to allow submission even with wrong answers (I did it by simplyfing the last if clause!).

**3) Most important, in my view, if answers are given in the gaps and the option "check_answers" is TRUE, the final data should present a boolean "accuracy" array with the comparison from answers supposed in the gaps and answers provided (or not) by the participant (No idea in how to do it! Please, if you can figure out in how to implement it, it would be great!).** 

4) It would be nice if the gaps accept an OR operator (e.g., % TV|television %). 

Also, I just try to save the data "answers" in a MySql VARCHAR, it just appears as 'Array'. Any idea in how to correctly catch the answers in the MySql?
Thanks a lot for all your attention and helping!

```
/**
 * jspsych-cloze
 * Philipp Sprengholz
 *
 * Plugin for displaying a cloze test and checking participants answers against a correct solution.
 *
 * documentation: docs.jspsych.org
 **/

jsPsych.plugins['cloze'] = (function () {

    var plugin = {};

    plugin.info = {
        name: 'cloze',
        description: '',
        parameters: {
            text: {
                type: jsPsych.plugins.parameterType.STRING,
                pretty_name: 'Cloze text',
                default: undefined,
                description: 'The cloze text to be displayed. Blanks are indicated by %% signs and automatically replaced by input fields. If there is a correct answer you want the system to check against, it must be typed between the two percentage signs (i.e. %solution%).'
            },
            button_text: {
                type: jsPsych.plugins.parameterType.STRING,
                pretty_name: 'Button text',
                default: 'OK',
                description: 'Text of the button participants have to press for finishing the cloze test.'
            },
            check_answers: {
                type: jsPsych.plugins.parameterType.BOOL,
                pretty_name: 'Check answers',
                default: false,
                description: 'Boolean value indicating if the answers given by participants should be compared against a correct solution given in the text (between % signs) after the button was clicked.'
            },
            mistake_fn: {
                type: jsPsych.plugins.parameterType.FUNCTION,
                pretty_name: 'Mistake function',
                default: function () {},
                description: 'Function called if check_answers is set to TRUE and there is a difference between the participants answers and the correct solution provided in the text.'
            }
        }
    };

    plugin.trial = function (display_element, trial) {

        var html = '<div class="cloze">';
        var elements = trial.text.split('%');
        var solutions = [];

        for (i=0; i<elements.length; i++)
        {
            if (i%2 === 0)
            {
                html += elements[i];
            }
            else
            {
                solutions.push(elements[i].trim());
                html += '<input type="text" id="input'+(solutions.length-1)+'" value="">';
            }
        }
        html += '</div>';
        
        display_element.innerHTML = html;
        
	var start_time = performance.now();
        
        var check = function() {

            var answers = [];
            var answers_correct = true;

            for (i=0; i<solutions.length; i++)
            {
                var field = document.getElementById('input'+i);
                answers.push(field.value.trim());
                
                if (trial.check_answers)
                {
                    if (answers[i] !== solutions[i])
                    {
                        field.style.color = 'red';
                        answers_correct = false;
                    }
                    else
                    {
                        field.style.color = 'black';
                    }
                }
            }
            
		var end_time = performance.now();
		rt = end_time - start_time;

		if (!trial.check_answers || (trial.check_answers && answers_correct))
            {
                var trial_data = {
                    'rt' : rt,
                    'answers' : answers
                };

                display_element.innerHTML = '';
                jsPsych.finishTrial(trial_data);
            }
            else
            {
                trial.mistake_fn();
            }

        };
                
		display_element.innerHTML += '<br><button class="jspsych-html-button-response-button" type="button" id="finish_cloze_button">'+trial.button_text+'</button>';
        display_element.querySelector('#finish_cloze_button').addEventListener('click', check);
    };

    return plugin;
})();
```

**Comment by jodeleeuw** [comment]
These are great ideas. If you want to start working towards a pull request that would be wonderful.

> I just try to save the data "answers" in a MySql VARCHAR, it just appears as 'Array'. Any idea in how to correctly catch the answers in the MySql?

We recommend that all data saved in jsPsych is a primitive type, like a number or character string. For arrays and objects, plugins generally convert these to JSON before storing them in the data.

```js
var trial_data = {
	'rt' : rt,
        'answers' : JSON.stringify(answers)
};
```

**Reply by gustavoestivalet** [reply]
Hello @jodeleeuw , thank you! It seems a nice project for this week, I have two other students applying cloze experiments and it will be useful.
I tried the stringify solution above, but it keeps appearing as 'Array' in MySql. Any other hint?
I followed the jsPsych webpage for PHP/MySql and it works perfectly for other experiments!
![image](https://user-images.githubusercontent.com/68119818/88574921-75f56e00-d019-11ea-8322-851ae05088b3.png)

Best,

**Reply by jodeleeuw** [reply]
Hi @gustavoestivalet . Just checking, but I think you resolved this issue in another thread right?

**Reply by gustavoestivalet** [reply]
Hi @jodeleeuw ! Yes, collation and Array storing perfectly solved (PHP update). Everything working good in localhost and webserver. I also started to use cognition.run, amazing! Not so much time to work in the changes in the cloze-plugin, but I'll keep you updated. Thank you for all!
